homeassistant/components/pvpc_hourly_pricing/sensor.py

```diff
@@ -0,0 +1,195 @@
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_NAME
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.event import (
+    async_track_point_in_time,
+    async_track_time_change,
+)
+from homeassistant.helpers.restore_state import RestoreEntity
+import homeassistant.util.dt as dt_util
+
+from .const import ATTR_TARIFF
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_PRICE = "price"
+ICON = "mdi:currency-eur"
+UNIT = "€/kWh"
+
+_DEFAULT_TIMEOUT = 10
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, config_entry: config_entries.ConfigEntry, async_add_entities
+):
+    """Set up the electricity price sensor from config_entry."""
+    name = config_entry.data[CONF_NAME]
+    pvpc_data_handler = PVPCData(
+        tariff=config_entry.data[ATTR_TARIFF],
+        local_timezone=hass.config.time_zone,
+        websession=async_get_clientsession(hass),
+        logger=_LOGGER,
+        timeout=_DEFAULT_TIMEOUT,
+    )
+    async_add_entities(
+        [ElecPriceSensor(name, config_entry.unique_id, pvpc_data_handler)], True
+    )
+
+
+class ElecPriceSensor(RestoreEntity):
+    """Class to hold the prices of electricity as a sensor."""
+
+    unit_of_measurement = UNIT
+    icon = ICON
+    should_poll = False
+
+    def __init__(self, name, unique_id, pvpc_data_handler):
+        """Initialize the sensor object."""
+        self._name = name
+        self._unique_id = unique_id
+        self._pvpc_data = pvpc_data_handler
+        self._num_retries = 0
+
+        self._init_done = False
+        self._hourly_tracker = None
+        self._price_tracker = None
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Cancel listeners for sensor updates."""
+        self._hourly_tracker()
+        self._price_tracker()
+
+    async def async_added_to_hass(self):
+        """Handle entity which will be added."""
+        await super().async_added_to_hass()
+        state = await self.async_get_last_state()
+        if state:
+            self._pvpc_data.state = state.state
+
+        # Update 'state' value in hour changes
+        self._hourly_tracker = async_track_time_change(
+            self.hass, self.async_update, second=[0], minute=[0]
+        )
+        # Update prices at random time, 2 times/hour (don't want to upset API)
+        random_minute = randint(1, 29)
+        mins_update = [random_minute, random_minute + 30]
+        self._price_tracker = async_track_time_change(
+            self.hass, self.async_update_prices, second=[0], minute=mins_update,
+        )
+        _LOGGER.debug(
+            "Setup of price sensor %s (%s) with tariff '%s', "
+            "updating prices each hour at %s min",
+            self.name,
+            self.entity_id,
+            self._pvpc_data.tariff,
+            mins_update,
+        )
+        await self.async_update_prices(dt_util.utcnow())
```

Comment:
It looks like async_update_prices can cause a `async_schedule_update_ha_state` if the source is unavailable so you could end up with two state updates.

Reply:
I'll review these. I think I remember there is some reason for the specific flow and the calls to `self.async_update_ha_state` with/wo force update. 

I'll take a look at the logic. 
FYI, async_update does no IO, just gets the price for the current hour. The request is made in the self.async_update_prices, which downloads daily packets of 24 prices.

Reply:
You were right! thanks!

The flow logic was outdated since I did the availability to meet silver requirements. I've simplified a lot the `async_update` now